### PR TITLE
Add a page to display all Recurrence rules

### PR DESCRIPTION
**Pull request by vladimir-8**
Screenshot 2024-10-31 at 16 04 00

includes:

- Recurrence rules ("recurring orders") created via the Delivery form
- Recurrence rules created via the Dispatch dashboard (possible options: Order+Delivery; Delivery without an Order/price; one/many Tasks without a Delivery)

### CODE REVIEW

File: src/Form/AddressBookType.php [hunk 1]
```diff
@@ -133,24 +133,29 @@ public function buildForm(FormBuilderInterface $builder, array $options)
 
             if (null !== $address) {
 
-                $addresses = $options['with_addresses'];
-                if (!is_array($addresses)) {
-                    if (is_callable([ $addresses, 'toArray' ])) {
-                        $addresses = $addresses->toArray();
+                $addressBook = $options['with_addresses'];
```
Comment (Atala) [inline src/Form/AddressBookType.php:136]: should be `addressesBook`? (plural)

File: src/Service/DeliveryManager.php [hunk 2]
```diff
@@ -146,4 +135,79 @@ public function setDefaults(Delivery $delivery)
...
+
+            if ($persist) {
+                $this->entityManager->persist($delivery);
+            }
+        }
+
+        return $delivery;
+    }
+
+    public function calculateRoute(TaskCollectionInterface $taskCollection): void
```
Comment (Atala) [inline src/Service/DeliveryManager.php:196]: wouldnt that be calculated automatically by `TaskListSubscriber` ?

File: templates/store/recurrence_rules.html.twig [hunk 3]
```diff
@@ -0,0 +1,24 @@
...
+
+{% block store_content %}
+
+  <h3>{{ 'stores.recurrence_rules'|trans }}</h3>
+
+  {% include '_partials/recurrence_rule/list.html.twig' with { items: recurrence_rules, empty_message: 'stores.recurrence_rules.empty_message' } %}
+
+{% endblock %}
+
+{#{% block styles %}#}
```
Comment (Atala) [inline templates/store/recurrence_rules.html.twig:18]: remove commented section

**Comment by Atala** [comment]
I made it crash :
- go to the dispatch and create a delivery rule with just 1 dropoff
- go to recurrence rules list page for the store

image

**Comment by Atala** [comment]
it does seem to me that the dropdown list to select a store when creating a recurrent rule in the dispatch does not display all the stores. is that expected? i can look deeper

**Comment by vladimir-8** [comment]
> in french breadcrumb it is still abonnements ("subscription")

Are you sure it was the right branch? Looks good on my machine:

Screenshot 2024-11-07 at 11 47 56

**Comment by vladimir-8** [comment]
> actually i thought the recurrence rules list page would be in the sidebar on "http://localhost/admin/deliveries" not on the store level. i guess admin users would like to see all the subscriptions for all his store. what do you think?

Yeah, I discussed it with Paul and we decided to put it on a store level for now. Maybe, at some point, we will make a global one as well

File: src/Form/AddressBookType.php (hunk 1 above)
Comment (vladimir-8) [reply src/Form/AddressBookType.php:136]: No, it looks correct: https://en.wikipedia.org/wiki/Address_book

File: src/Service/DeliveryManager.php (hunk 2 above)
Comment (vladimir-8) [reply src/Service/DeliveryManager.php:196]: `TaskListSubscriber` is triggered only when an Entity is persisted. On the recurrence rules page, I only create  placeholders for Orders/Deliveries/Tasks, but don't persist them

**Comment by vladimir-8** [comment]
> it does seem to me that the dropdown list to select a store when creating a recurrent rule in the dispatch does not display all the stores. is that expected? i can look deeper

For me, it displays everything. I'd say don't spend time on it, as we want to migrate away from that flow

**Comment by vladimir-8** [comment]
> I made it crash :
> 
>     go to the dispatch and create a delivery rule with just 1 dropoff
>     go to recurrence rules list page for the store

It seems you were not running the latest version locally. The code in your stack trace is different from the last commit in this branch.

Now it handles both one pickup or one dropoff:

Screenshot 2024-11-07 at 13 10 40

**Review by Atala — APPROVED** [review]
thanks for your answer, i may have misstest the thing

just a short comment, in the translation string "recurrence_rules.table.generate_orders" the text should be "Generate tasks/orders" or even "Generate orders (as are getting rid of tasks) (or something like this)

i updated a french string for more clarity

**Pull request merged**